policy.go

```diff
@@ -659,13 +741,32 @@ func (p *policyPlugin) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dn
 		}
 	}
 
-	// validate domain name (validation #1)
-	err = p.validate(ah, "")
-	if err != nil {
-		status = dns.RcodeServerFailure
-		goto Exit
-	}
+	if p.rule.keyword != "" {
+		result, err := p.evaluateExp(p.rule.params, p.rule.keyword, w, r)
+		if err != nil {
+			resolveFailed = true
+			goto Exit
+		}
+		if result == true {
+			if p.rule.keyword == "permit" {
+				ah.action = typeAllow
 
+			} else if p.rule.keyword == "deny" {
+				ah.action = typeRefuse
+			}
+		}
+		if result == false {
+			ah.action = typeRefuse
+		}
+	}
```

Comment:
The above block L744-L769 should be a method of the policyRule object, which returns the result of the evaluation.